File: deno-debugger/scripts/cdp_client.ts
/**
 * Chrome DevTools Protocol (CDP) client for connecting to Deno's V8 Inspector.
 *
 * Provides high-level interface for debugging operations:
 * - Connection management
 * - Breakpoint control
 * - Execution flow (resume, pause, step)
 * - Expression evaluation
 * - Call frame and scope inspection
 * - Heap snapshot capture
 * - CPU profiling
 */

import type {
  BreakpointLocation,
  CallFrame,
  CDPEvent,
  CDPRequest,
  CDPResponse,
  CDPTarget,
  CPUProfileData,
  RemoteObject,
} from "./types.ts";

export interface Breakpoint {
  breakpointId: string;
  location: BreakpointLocation;
}

type EventHandler = (params: Record<string, unknown>) => void | Promise<void>;

export class CDPClient {
  private host: string;
  private port: number;
  private ws: WebSocket | null = null;
  private nextId = 1;
  private pendingRequests = new Map<number, {
    resolve: (value: Record<string, unknown>) => void;
    reject: (error: Error) => void;
  }>();
  private eventHandlers = new Map<string, EventHandler[]>();
  public paused = false;
  public callFrames: CallFrame[] = [];
  public runtimeInfo: {
    isDeno: boolean;
    isNode: boolean;
    description: string;
    title: string;
  } | null = null;

  constructor(host = "127.0.0.1", port = 9229) {
    this.host = host;
    this.port = port;
  }

  /**
   * Establish WebSocket connection to Deno/Node inspector.
   */
  async connect(): Promise<CDPClient> {
    // First, get the WebSocket debugger URL
    const response = await fetch(`http://${this.host}:${this.port}/json`);
    const targets = await response.json() as CDPTarget[];

    if (!targets || targets.length === 0) {
      throw new Error("No debugger targets found");
    }

    const target = targets[0];
    const wsUrl = target.webSocketDebuggerUrl;

    // Detect runtime from target info
    const description = (target.description || "").toLowerCase();
    const title = (target.title || "").toLowerCase();
    this.runtimeInfo = {
      isDeno: description.includes("deno") || title.includes("deno"),
      isNode: description.includes("node") || title.includes("node"),
      description: target.description || "",
      title: target.title || "",
    };

    // Connect to WebSocket
    this.ws = new WebSocket(wsUrl);

    // Set up event handlers
    return new Promise((resolve, reject) => {
      if (!this.ws) {
        reject(new Error("WebSocket not initialized"));
        return;
      }

      this.ws.onopen = () => {
        this.startMessageHandler();
        // Give message handler a moment to start
        setTimeout(() => resolve(this), 100);
      };

      this.ws.onerror = (event) => {
        reject(new Error(`WebSocket error: ${event}`));
      };

      this.ws.onclose = () => {
        // Connection closed
      };
    });
  }

  /**
   * Handle incoming messages from CDP.
   */
  private startMessageHandler(): void {
    if (!this.ws) return;

    this.ws.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data as string);

        // Handle responses to our requests
        if ("id" in data) {
          const response = data as CDPResponse;
          const pending = this.pendingRequests.get(response.id);
          if (pending) {
            this.pendingRequests.delete(response.id);
            if (response.error) {
              pending.reject(new Error(response.error.message));
            } else {
              pending.resolve(response.result || {});
            }
          }
        } // Handle events
        else if ("method" in data) {
          const event = data as CDPEvent;
          const method = event.method;
          const params = event.params || {};

          // Built-in event handling
          if (method === "Debugger.paused") {
            this.paused = true;
            this.callFrames = (params.callFrames as CallFrame[]) || [];
          } else if (method === "Debugger.resumed") {
            this.paused = false;
            this.callFrames = [];
          }

          // Notify registered handlers
          const handlers = this.eventHandlers.get(method);
          if (handlers) {
            for (const handler of handlers) {
              // Run handler async
              Promise.resolve(handler(params)).catch((err) => {
                console.error(`Event handler error for ${method}:`, err);
              });
            }
          }
        }
      } catch (err) {
        console.error("CDP message handler error:", err);
      }
    };
  }

  /**
   * Send a CDP command and wait for response.
   */
  async sendCommand(
    method: string,
    params?: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error("WebSocket not connected");
    }

    const msgId = this.nextId++;
    const message: CDPRequest = {
      id: msgId,
      method,
    };

    if (params) {
      message.params = params;
    }

    // Create promise for response
    const responsePromise = new Promise<Record<string, unknown>>((resolve, reject) => {
      this.pendingRequests.set(msgId, { resolve, reject });
    });

    // Send message
    this.ws.send(JSON.stringify(message));

    // Wait for response with timeout
    const timeoutPromise = new Promise<never>((_, reject) => {
      setTimeout(() => reject(new Error(`Command timeout: ${method}`)), 30000);
    });

    return await Promise.race([responsePromise, timeoutPromise]);
  }

  /**
   * Register an event handler.
   */
  onEvent(eventName: string, handler: EventHandler): void {
    if (!this.eventHandlers.has(eventName)) {
      this.eventHandlers.set(eventName, []);
    }
    this.eventHandlers.get(eventName)!.push(handler);
  }

  // ============================================================================
  // Debugger Domain
  // ============================================================================

  async enableDebugger(): Promise<void> {
    await this.sendCommand("Debugger.enable");
    // Also enable Runtime for evaluation
    await this.sendCommand("Runtime.enable");
  }

  async disableDebugger(): Promise<void> {
    await this.sendCommand("Debugger.disable");
  }

  async setBreakpoint(
    url: string,
    line: number,
    column = 0,
    condition?: string,
  ): Promise<Breakpoint> {
    const params: Record<string, unknown> = {
      location: { scriptUrl: url, lineNumber: line, columnNumber: column },
    };

    if (condition) {
      params.condition = condition;
    }

    const result = await this.sendCommand("Debugger.setBreakpoint", params);

    return {
      breakpointId: result.breakpointId as string,
      location: result.actualLocation as BreakpointLocation,
    };
  }

  async setBreakpointByUrl(
    urlRegex: string,
    line: number,
    column = 0,
    condition?: string,
  ): Promise<string> {
    const params: Record<string, unknown> = {
      urlRegex,
      lineNumber: line,
      columnNumber: column,
    };

    if (condition) {
      params.condition = condition;
    }

    const result = await this.sendCommand("Debugger.setBreakpointByUrl", params);
    return result.breakpointId as string;
  }

  async removeBreakpoint(breakpointId: string): Promise<void> {
    await this.sendCommand("Debugger.removeBreakpoint", { breakpointId });
  }

  async pause(): Promise<void> {
    await this.sendCommand("Debugger.pause");
  }

  async resume(): Promise<void> {
    await this.sendCommand("Debugger.resume");
  }

  async stepOver(): Promise<void> {
    await this.sendCommand("Debugger.stepOver");
  }

  async stepInto(): Promise<void> {
    await this.sendCommand("Debugger.stepInto");
  }

  async stepOut(): Promise<void> {
    await this.sendCommand("Debugger.stepOut");
  }

  async setPauseOnExceptions(state: "none" | "uncaught" | "all" = "none"): Promise<void> {
    await this.sendCommand("Debugger.setPauseOnExceptions", { state });
  }

  getCallFrames(): CallFrame[] {
    return this.callFrames;
  }

  async evaluate(
    expression: string,
    callFrameId?: string,
    contextId?: number,
  ): Promise<RemoteObject> {
    let result: Record<string, unknown>;

    if (callFrameId) {
      // Evaluate on call frame (when paused)
      result = await this.sendCommand("Debugger.evaluateOnCallFrame", {
        callFrameId,
        expression,
      });
    } else {
      // Evaluate in runtime context
      const params: Record<string, unknown> = { expression };
      if (contextId !== undefined) {
        params.contextId = contextId;
      }
      result = await this.sendCommand("Runtime.evaluate", params);
    }

    if (result.exceptionDetails) {
      throw new Error(`Evaluation error: ${JSON.stringify(result.exceptionDetails)}`);
    }

    return result.result as RemoteObject;
  }

  async getProperties(objectId: string): Promise<Array<Record<string, unknown>>> {
    const result = await this.sendCommand("Runtime.getProperties", {
      objectId,
      ownProperties: true,
    });
    return (result.result as Array<Record<string, unknown>>) || [];
  }

  async getScopeVariables(callFrameId: string): Promise<Record<string, unknown>> {
    const frame = this.callFrames.find((f) => f.callFrameId === callFrameId);
    if (!frame) {
      return {};
    }

    const variables: Record<string, unknown> = {};

    // Get variables from each scope chain
    for (const scope of frame.scopeChain || []) {
      const scopeObj = scope.object;
      if (scopeObj.objectId) {
        const props = await this.getProperties(scopeObj.objectId);
        for (const prop of props) {
          if (prop.name) {
            variables[prop.name as string] = prop.value;
          }
        }
      }
    }

    return variables;
  }

  // ============================================================================
  // Heap Profiler Domain
  // ============================================================================

  async enableHeapProfiler(): Promise<void> {
    await this.sendCommand("HeapProfiler.enable");
  }

  async takeHeapSnapshot(reportProgress = false): Promise<string> {
    const chunks: string[] = [];
    let progressDone = false;

    const chunkHandler = (params: Record<string, unknown>) => {
      if (params.chunk) {
        chunks.push(params.chunk as string);
      }
    };

    const progressHandler = (params: Record<string, unknown>) => {
      const done = params.done as number || 0;
      const total = params.total as number || 0;

      // Print progress if requested
      if (reportProgress && done % 20000 === 0) {
        console.log(`  Heap snapshot progress: ${done.toLocaleString()}/${total.toLocaleString()}`);
      }

      // When finished is true, snapshot is complete
      if (params.finished) {
        progressDone = true;
      }
    };

    // Register event handlers BEFORE enabling heap profiler
    this.onEvent("HeapProfiler.addHeapSnapshotChunk", chunkHandler);

    if (reportProgress) {
      this.onEvent("HeapProfiler.reportHeapSnapshotProgress", progressHandler);
    }

    // Now enable heap profiler
    await this.enableHeapProfiler();

    // Request snapshot
    try {
      // Start snapshot capture
      const snapshotPromise = this.sendCommand("HeapProfiler.takeHeapSnapshot", {
        reportProgress,
      });

      // Wait for completion
      if (reportProgress) {
        // Wait for progress indicator
        let waited = 0;
        while (!progressDone && waited < 30000) {
          await new Promise((resolve) => setTimeout(resolve, 100));
          waited += 100;
        }
      } else {
        // Wait a bit for chunks to arrive
        await new Promise((resolve) => setTimeout(resolve, 5000));
      }

      // Wait for command completion
      await snapshotPromise;
    } catch (err) {
      // Continue even if command times out - we have the chunks
      console.error("Snapshot capture error (may be OK):", err);
    }

    return chunks.join("");
  }

  // ============================================================================
  // Profiler Domain (CPU profiling)
  // ============================================================================

  async enableProfiler(): Promise<void> {
    await this.sendCommand("Profiler.enable");
  }

  async startProfiling(): Promise<void> {
    await this.enableProfiler();
    await this.sendCommand("Profiler.start");
  }

  async stopProfiling(): Promise<CPUProfileData> {
    const result = await this.sendCommand("Profiler.stop");
    return (result.profile as CPUProfileData) || {} as CPUProfileData;
  }

  close(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}

// ============================================================================
// CLI Usage
// ============================================================================

if (import.meta.main) {
  console.log("CDP Client - Example Usage");
  console.log("===========================");
  console.log();
  console.log("// Connect to Deno (launched with --inspect)");
  console.log("const client = new CDPClient('127.0.0.1', 9229);");
  console.log("await client.connect();");
  console.log("await client.enableDebugger();");
  console.log();
  console.log("// Set breakpoint");
  console.log("const bp = await client.setBreakpoint('file:///path/to/app.ts', 42);");
  console.log();
  console.log("// Resume execution");
  console.log("await client.resume();");
  console.log();
  console.log("// When paused, inspect");
  console.log("const frames = client.getCallFrames();");
  console.log("const vars = await client.getScopeVariables(frames[0].callFrameId);");
  console.log();
  console.log("// Evaluate expression");
  console.log("const result = await client.evaluate('myVariable', frames[0].callFrameId);");
  console.log();
  console.log("// Take heap snapshot");
  console.log("const snapshot = await client.takeHeapSnapshot();");
  console.log("await Deno.writeTextFile('snapshot.heapsnapshot', snapshot);");
  console.log();
  console.log("// CPU profiling");
  console.log("await client.startProfiling();");
  console.log("// ... run code ...");
  console.log("const profile = await client.stopProfiling();");
  console.log("await Deno.writeTextFile('profile.cpuprofile', JSON.stringify(profile));");
}

File: skill/scripts/cdp_client.py
"""
Chrome DevTools Protocol (CDP) client for connecting to Deno's V8 Inspector.

Provides high-level interface for debugging operations:
- Connection management
- Breakpoint control
- Execution flow (resume, pause, step)
- Expression evaluation
- Call frame and scope inspection
- Heap snapshot capture
- CPU profiling
"""

import asyncio
import json
import websockets
from typing import Any, Dict, List, Optional, Callable
from dataclasses import dataclass


@dataclass
class Breakpoint:
    """Represents a breakpoint in source code."""
    breakpoint_id: str
    location: Dict[str, Any]


class CDPClient:
    """
    Chrome DevTools Protocol client for Deno debugging.

    Usage:
        client = CDPClient('127.0.0.1', 9229)
        await client.connect()
        await client.enable_debugger()
        bp = await client.set_breakpoint('file:///path/to/file.ts', 42)
        await client.resume()
    """

    def __init__(self, host: str = '127.0.0.1', port: int = 9229):
        self.host = host
        self.port = port
        self.ws = None
        self.next_id = 1
        self.pending_requests: Dict[int, asyncio.Future] = {}
        self.event_handlers: Dict[str, List[Callable]] = {}
        self.paused = False
        self.call_frames = []
        self.runtime_info = None  # Will store Deno vs Node detection

    async def connect(self):
        """Establish WebSocket connection to Deno/Node inspector."""
        # First, get the WebSocket debugger URL
        import aiohttp
        async with aiohttp.ClientSession() as session:
            async with session.get(f'http://{self.host}:{self.port}/json') as resp:
                targets = await resp.json()
                if not targets:
                    raise Exception("No debugger targets found")

                target = targets[0]
                ws_url = target['webSocketDebuggerUrl']

                # Detect runtime from target info
                # Deno includes "Deno" in description, Node includes "node"
                description = target.get('description', '').lower()
                title = target.get('title', '').lower()
                self.runtime_info = {
                    'is_deno': 'deno' in description or 'deno' in title,
                    'is_node': 'node' in description or 'node' in title,
                    'description': target.get('description', ''),
                    'title': target.get('title', '')
                }

        # Connect to WebSocket
        self.ws = await websockets.connect(ws_url)

        # Start message handler
        asyncio.create_task(self._message_handler())

        return self

    async def _message_handler(self):
        """Handle incoming messages from CDP."""
        try:
            async for message in self.ws:
                data = json.loads(message)

                # Handle responses to our requests
                if 'id' in data:
                    msg_id = data['id']
                    if msg_id in self.pending_requests:
                        future = self.pending_requests.pop(msg_id)
                        if 'error' in data:
                            future.set_exception(Exception(data['error']['message']))
                        else:
                            future.set_result(data.get('result', {}))

                # Handle events
                elif 'method' in data:
                    method = data['method']
                    params = data.get('params', {})

                    # Debug: log all events (comment out for production)
                    if 'HeapProfiler' in method:
                        print(f"    [DEBUG] Event: {method}")

                    # Built-in event handling
                    if method == 'Debugger.paused':
                        self.paused = True
                        self.call_frames = params.get('callFrames', [])
                    elif method == 'Debugger.resumed':
                        self.paused = False
                        self.call_frames = []

                    # Notify registered handlers
                    if method in self.event_handlers:
                        for handler in self.event_handlers[method]:
                            asyncio.create_task(handler(params))

        except websockets.exceptions.ConnectionClosed:
            pass

    async def send_command(self, method: str, params: Optional[Dict] = None) -> Dict[str, Any]:
        """
        Send a CDP command and wait for response.

        Args:
            method: CDP method name (e.g., 'Debugger.resume')
            params: Optional parameters dict

        Returns:
            Response result dict
        """
        msg_id = self.next_id
        self.next_id += 1

        message = {
            'id': msg_id,
            'method': method,
        }
        if params:
            message['params'] = params

        # Create future for response
        future = asyncio.Future()
        self.pending_requests[msg_id] = future

        # Send message
        await self.ws.send(json.dumps(message))

        # Wait for response
        return await future

    def on_event(self, event_name: str, handler: Callable):
        """
        Register an event handler.

        Args:
            event_name: CDP event name (e.g., 'Debugger.paused')
            handler: Async function to call with event params
        """
        if event_name not in self.event_handlers:
            self.event_handlers[event_name] = []
        self.event_handlers[event_name].append(handler)

    # Debugger Domain

    async def enable_debugger(self):
        """Enable the Debugger domain."""
        await self.send_command('Debugger.enable')
        # Also enable Runtime for evaluation
        await self.send_command('Runtime.enable')

    async def disable_debugger(self):
        """Disable the Debugger domain."""
        await self.send_command('Debugger.disable')

    async def set_breakpoint(
        self,
        url: str,
        line: int,
        column: int = 0,
        condition: Optional[str] = None
    ) -> Breakpoint:
        """
        Set a breakpoint in source code.

        Args:
            url: File URL (e.g., 'file:///path/to/file.ts')
            line: Line number (0-indexed)
            column: Column number (0-indexed)
            condition: Optional condition expression

        Returns:
            Breakpoint object
        """
        params = {
            'location': {
                'scriptUrl': url,
                'lineNumber': line,
                'columnNumber': column
            }
        }
        if condition:
            params['condition'] = condition

        result = await self.send_command('Debugger.setBreakpoint', params)

        return Breakpoint(
            breakpoint_id=result['breakpointId'],
            location=result['actualLocation']
        )

    async def set_breakpoint_by_url(
        self,
        url_regex: str,
        line: int,
        column: int = 0,
        condition: Optional[str] = None
    ) -> str:
        """
        Set breakpoint by URL pattern (useful when script hasn't loaded yet).

        Returns:
            Breakpoint ID
        """
        params = {
            'urlRegex': url_regex,
            'lineNumber': line,
            'columnNumber': column
        }
        if condition:
            params['condition'] = condition

        result = await self.send_command('Debugger.setBreakpointByUrl', params)
        return result['breakpointId']

    async def remove_breakpoint(self, breakpoint_id: str):
        """Remove a breakpoint."""
        await self.send_command('Debugger.removeBreakpoint', {
            'breakpointId': breakpoint_id
        })

    async def pause(self):
        """Pause execution."""
        await self.send_command('Debugger.pause')

    async def resume(self):
        """Resume execution."""
        await self.send_command('Debugger.resume')

    async def step_over(self):
        """Step over to next statement."""
        await self.send_command('Debugger.stepOver')

    async def step_into(self):
        """Step into function call."""
        await self.send_command('Debugger.stepInto')

    async def step_out(self):
        """Step out of current function."""
        await self.send_command('Debugger.stepOut')

    async def set_pause_on_exceptions(self, state: str = 'none'):
        """
        Set pause behavior for exceptions.

        Args:
            state: 'none', 'uncaught', or 'all'
        """
        await self.send_command('Debugger.setPauseOnExceptions', {
            'state': state
        })

    async def get_call_frames(self) -> List[Dict[str, Any]]:
        """
        Get current call frames (when paused).

        Returns:
            List of call frame objects
        """
        return self.call_frames

    async def evaluate(
        self,
        expression: str,
        call_frame_id: Optional[str] = None,
        context_id: Optional[int] = None
    ) -> Any:
        """
        Evaluate an expression.

        Args:
            expression: JavaScript/TypeScript expression
            call_frame_id: Evaluate in context of this call frame (when paused)
            context_id: Execution context ID (for global evaluation)

        Returns:
            Evaluation result
        """
        if call_frame_id:
            # Evaluate on call frame (when paused)
            result = await self.send_command('Debugger.evaluateOnCallFrame', {
                'callFrameId': call_frame_id,
                'expression': expression
            })
        else:
            # Evaluate in runtime context
            params = {'expression': expression}
            if context_id:
                params['contextId'] = context_id
            result = await self.send_command('Runtime.evaluate', params)

        if result.get('exceptionDetails'):
            raise Exception(f"Evaluation error: {result['exceptionDetails']}")

        return result['result']

    async def get_properties(self, object_id: str) -> List[Dict[str, Any]]:
        """
        Get properties of an object.

        Args:
            object_id: Remote object ID

        Returns:
            List of property descriptors
        """
        result = await self.send_command('Runtime.getProperties', {
            'objectId': object_id,
            'ownProperties': True
        })
        return result.get('result', [])

    async def get_scope_variables(self, call_frame_id: str) -> Dict[str, Any]:
        """
        Get all variables in scope for a call frame.

        Args:
            call_frame_id: Call frame ID

        Returns:
            Dict mapping variable names to values
        """
        frame = None
        for f in self.call_frames:
            if f['callFrameId'] == call_frame_id:
                frame = f
                break

        if not frame:
            return {}

        variables = {}

        # Get variables from each scope chain
        for scope in frame.get('scopeChain', []):
            scope_obj = scope['object']
            if 'objectId' in scope_obj:
                props = await self.get_properties(scope_obj['objectId'])
                for prop in props:
                    if prop.get('name'):
                        variables[prop['name']] = prop.get('value')

        return variables

    # Heap Profiler Domain

    async def enable_heap_profiler(self):
        """Enable heap profiler."""
        await self.send_command('HeapProfiler.enable')

    async def take_heap_snapshot(self, report_progress: bool = False) -> str:
        """
        Take a heap snapshot.

        ⚠️  KNOWN ISSUE: Deno's V8 inspector does NOT send heap snapshot chunks.
        This method will return empty string when connected to Deno.

        Workaround: Use Chrome DevTools UI to manually capture heap snapshots
        and export them as .heapsnapshot files for analysis.

        See docs/DENO_HEAP_SNAPSHOT_BUG.md for details.

        Returns:
            Heap snapshot as JSON string (can be large!)
            Returns empty string if connected to Deno due to known bug.
        """
        # Warn if connected to Deno
        if self.runtime_info and self.runtime_info.get('is_deno'):
            import warnings
            warnings.warn(
                "\n"
                "⚠️  HEAP SNAPSHOT LIMITATION: Deno's V8 inspector does not send heap snapshot chunks.\n"
                "   This is a known Deno bug. takeHeapSnapshot will return empty data.\n"
                "\n"
                "   Workaround: Use Chrome DevTools UI to manually capture heap snapshots:\n"
                "   1. Open chrome://inspect in Chrome\n"
                "   2. Click 'inspect' on your Deno process\n"
                "   3. Go to Memory tab\n"
                "   4. Click 'Take snapshot'\n"
                "   5. Right-click snapshot and 'Save as...'\n"
                "   6. Load the .heapsnapshot file with our HeapSnapshot class\n"
                "\n"
                "   See docs/DENO_HEAP_SNAPSHOT_BUG.md for full details.\n",
                UserWarning,
                stacklevel=2
            )

        await self.enable_heap_profiler()

        chunks = []
        done_event = asyncio.Event()
        progress_done = asyncio.Event()

        async def chunk_handler(params):
            if 'chunk' in params:
                chunks.append(params['chunk'])
                # Debug: print chunk progress
                if len(chunks) % 100 == 0:
                    print(f"    Received {len(chunks)} chunks...")

        async def progress_handler(params):
            # Debug: print progress
            done = params.get('done', 0)
            total = params.get('total', 0)
            print(f"    Progress: {done}/{total}")

            # When finished is True, snapshot is complete
            if params.get('finished'):
                print(f"    Snapshot complete!")
                progress_done.set()

        # Listen for heap snapshot chunks
        self.on_event('HeapProfiler.addHeapSnapshotChunk', chunk_handler)

        # Listen for progress (if reportProgress is True)
        if report_progress:
            self.on_event('HeapProfiler.reportHeapSnapshotProgress', progress_handler)

        # Request snapshot (don't await - command might not return until acknowledged)
        try:
            print(f"    Sending takeHeapSnapshot command...")
            # Send command in background - we'll wait for events instead
            asyncio.create_task(self.send_command('HeapProfiler.takeHeapSnapshot', {
                'reportProgress': report_progress
            }))

            # Wait for progress to indicate completion (if we're tracking progress)
            # Otherwise wait a bit for chunks
            if report_progress:
                print(f"    Waiting for progress completion...")
                await asyncio.wait_for(progress_done.wait(), timeout=30)
            else:
                # Wait for chunks to arrive - increase timeout for large heaps
                print(f"    Waiting for chunks...")
                await asyncio.sleep(5.0)

            print(f"    Got {len(chunks)} total chunks")

        except asyncio.TimeoutError:
            print(f"  Warning: Heap snapshot timed out, got {len(chunks)} chunks")

        finally:
            # Clean up event handlers
            if 'HeapProfiler.addHeapSnapshotChunk' in self.event_handlers:
                if chunk_handler in self.event_handlers['HeapProfiler.addHeapSnapshotChunk']:
                    self.event_handlers['HeapProfiler.addHeapSnapshotChunk'].remove(chunk_handler)
            if report_progress and 'HeapProfiler.reportHeapSnapshotProgress' in self.event_handlers:
                if progress_handler in self.event_handlers['HeapProfiler.reportHeapSnapshotProgress']:
                    self.event_handlers['HeapProfiler.reportHeapSnapshotProgress'].remove(progress_handler)

        return ''.join(chunks)

    # Profiler Domain (CPU profiling)

    async def enable_profiler(self):
        """Enable CPU profiler."""
        await self.send_command('Profiler.enable')

    async def start_profiling(self):
        """Start CPU profiling."""
        await self.enable_profiler()
        await self.send_command('Profiler.start')

    async def stop_profiling(self) -> Dict[str, Any]:
        """
        Stop CPU profiling and return profile.

        Returns:
            CPU profile data
        """
        result = await self.send_command('Profiler.stop')
        return result.get('profile', {})

    async def close(self):
        """Close the connection."""
        if self.ws:
            await self.ws.close()


# Synchronous wrapper for easier use in REPL/scripts
class CDPClientSync:
    """
    Synchronous wrapper around CDPClient for easier interactive use.

    Usage:
        client = CDPClientSync('127.0.0.1', 9229)
        client.enable_debugger()
        bp = client.set_breakpoint('file:///app.ts', 42)
        client.resume()
    """

    def __init__(self, host: str = '127.0.0.1', port: int = 9229):
        self.client = CDPClient(host, port)
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(self.loop)
        self.loop.run_until_complete(self.client.connect())

    def __getattr__(self, name):
        """Proxy all methods to async client, running them synchronously."""
        attr = getattr(self.client, name)
        if asyncio.iscoroutinefunction(attr):
            def sync_wrapper(*args, **kwargs):
                return self.loop.run_until_complete(attr(*args, **kwargs))
            return sync_wrapper
        return attr

    def close(self):
        """Close connection and event loop."""
        self.loop.run_until_complete(self.client.close())
        self.loop.close()


if __name__ == '__main__':
    # Example usage
    print("CDP Client - Example Usage")
    print("===========================")
    print()
    print("# Connect to Deno (launched with --inspect)")
    print("client = CDPClientSync('127.0.0.1', 9229)")
    print("client.enable_debugger()")
    print()
    print("# Set breakpoint")
    print("bp = client.set_breakpoint('file:///path/to/app.ts', 42)")
    print()
    print("# Resume execution")
    print("client.resume()")
    print()
    print("# When paused, inspect")
    print("frames = client.get_call_frames()")
    print("vars = client.get_scope_variables(frames[0]['callFrameId'])")
    print()
    print("# Evaluate expression")
    print("result = client.evaluate('myVariable', frames[0]['callFrameId'])")
